network_topology_strategy: reallocate_tablets: support for rf changes

bhalevy opened the pull request:
This series provides a reallocate_tablets function, that's initially called by allocate_tablets_for_new_table.
The new allocation implementation is independent of vnodes/token ownership.
Rather than using the natural_endpoints_tracker, it implements its own tracking
based on dc/rack load (== number of replicas in rack), with the additional benefit
that tablet allocation will balance the allocation across racks, using a heap structure,
similar to the one we use to balance tablet allocation across shards in each node.

reallocate_tablets may also be called with an optional parameter pointing the the current tablet_map.
In this case the function either allocates more tablet replicas in datacenters for which the replication factor was increased,
or it will deallocate tablet replicas from datacenters for which replication factor was decreased.

The NetworkTopologyStrategy_tablets_test unit test was extended to cover replication factor changes.

bhalevy commented:
Refs https://github.com/scylladb/scylladb/pull/17116

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7470/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7470/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8569/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/537/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7470/)
#### Failed Tests (5/29732):
- [test_simple_kill_new_node_while_bootstrapping_with_parallel_writes_in_multidc](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7470/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest_with_topology_changes___test_simple_kill_new_node_while_bootstrapping_with_parallel_writes_in_multidc) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_kill_new_node_while_bootstrapping_with_parallel_writes_in_multidc)
- [test_recover_stuck_raft_recovery](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7470/testReport/junit/%28root%29/test_raft_recovery_stuck/Tests___Unit_Tests___test_recover_stuck_raft_recovery) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_recover_stuck_raft_recovery)
- [test_replace_ignore_nodes](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7470/testReport/junit/%28root%29/test_replace_ignore_nodes/Tests___Unit_Tests___test_replace_ignore_nodes) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replace_ignore_nodes)
- [topology_custom.test_raft_recovery_stuck.dev.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7470/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_custom_test_raft_recovery_stuck_dev_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_custom.test_raft_recovery_stuck.dev.2)
- [topology_custom.test_replace_ignore_nodes.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7470/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_custom_test_replace_ignore_nodes_dev_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_custom.test_replace_ignore_nodes.dev.3)
#### Build Details:
- Duration: 2 hr 5 min
- Builder: spider2.cloudius-systems.com

bhalevy commented:
All failures in ci aren't related to this change.  rerunning ci

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7478/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7478/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8574/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/542/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7478/)
#### Build Details:
- Duration: 2 hr 5 min
- Builder: spider5.cloudius-systems.com

xemul commented:
> Refs #17116

@bhalevy , please expand the "refs". How exactly this change is going to facilitate #17116

avikivity commented on test/boost/network_topology_strategy_test.cc:
```diff
@@ -132,7 +133,8 @@ void endpoints_check(
     }
 
     for (auto&& [dc, rf] : dc_rf) {
```
This reminds me: when dropping rf, if a rack is down (or even just a node in the rack), drop that rack preferentially. This allows easier recovery from the complete loss of a rack.

bhalevy replied:
I discussed dropping replicas with @tgrabiec and we came to the conclusion that we want to always drop replicas in reverse ordinal order so not to break replica pairing for MVs.

With only a node down in a rack, the base table might decide to drop this replica, while the view places a replica on a live node in that rack and will reach a different solution, so you proposal makes sense only when the whole rack is down.

Maybe we should consider an extension to the network-topology-strategy options so one could specify rf per rack (e.g. `{'dc/rack': 'rf', ...}` and then it could be set to 0 in case the rack is to be removed.

Cc @nyh

xemul commented on locator/network_topology_strategy.cc:
```diff
@@ -336,39 +335,128 @@ future<tablet_map> network_topology_strategy::allocate_tablets_for_new_table(sch
+    return reallocate_tablets(std::move(s), std::move(tm), tablet_count, nullptr);
+}
+
+future<tablet_map> network_topology_strategy::reallocate_tablets(schema_ptr s, token_metadata_ptr tm, size_t tablet_count, const tablet_map* cur_tablets) const {
```
THe cur_tablets is non-null for tests only, why not make _tests_ pass here the `tablet_map(tablet_count)`?

bhalevy replied:
The idea is for RF changes to use this interface and passing a pointer to the current tablet_map.
I'm not sure what exactly you're suggesting.

bhalevy replied:
good idea

xemul replied:
> The idea is for RF changes to use this interface and passing a pointer to the current tablet_map.

Ah, OK. Then my comment is irrelevant

bhalevy replied:
done in v2

xemul commented on locator/network_topology_strategy.cc:
```diff
@@ -336,39 +335,128 @@ future<tablet_map> network_topology_strategy::allocate_tablets_for_new_table(sch
-        for (auto&& ep : tracker.replicas()) {
-            replicas.emplace_back(tablet_replica{ep, load.next_shard(ep)});
+            if (dc_rf > dr_load.node_count) {
+                auto cmp = [] (const rack_load& a, const rack_load& b) {
```
Can struct rack_load_cmp (introduced in this patch up above) be reused?

bhalevy replied:
yes sorry, it was used initially, then I want to use a function, thinking that in the drop-replicas case we could use a descending order cmp function for the heap so to evict from the most loaded racks, but then I realized we want to drop replicas in reverse ordinal order to maintain base/view replica pairing, and forgot to reuse rack_load_cmp.
Will do in next version.

bhalevy replied:
in v2 I dropped the heap, but still use a similar node_load_cmp struct

xemul commented on locator/network_topology_strategy.cc:
```diff
@@ -336,39 +335,128 @@ future<tablet_map> network_topology_strategy::allocate_tablets_for_new_table(sch
+            replicas.reserve(_rep_factor);
         }
+        for (const auto& [dc, dc_rf] : _dc_rep_factor) {
+            auto& dr_load = dc_rack_load_map[dc];
```
For the sake of readability improvement -- does it make sense to wrap the hole body of this loop into reallocate_tablets_in_dc()?

bhalevy replied:
yes, we can refactor the loop body out into an internal function.

bhalevy replied:
refactored in v2

avikivity commented on locator/network_topology_strategy.cc:
```diff
@@ -336,39 +335,128 @@ future<tablet_map> network_topology_strategy::allocate_tablets_for_new_table(sch
+            if (dc_rf > dr_load.node_count) {
+                auto cmp = [] (const rack_load& a, const rack_load& b) {
+                    return a.load > b.load;
+                };
```
This is rack_load_cmp.

avikivity commented on locator/network_topology_strategy.cc:
```diff
@@ -336,39 +335,128 @@ future<tablet_map> network_topology_strategy::allocate_tablets_for_new_table(sch
+                        continue;
+                    }
+                    const auto& host_id = rack.back();
+                    replicas.emplace_back(tablet_replica{host_id, load.next_shard(host_id)});
```
We need to keep racks in the same order in the replica list. Migrations only happen in the same ordinal position within the replica list, and this keeps  the migration within the rack.

So we need to special-case dc_rf == all_dc_racks.size(), and in that special case, have a 1:1 relationship between the ordinal position in replicas and the ordinal position in all_dc_racks.

avikivity replied:
We also need to document
 - when adding a datacenter, add it all at once
 - prefer having rack replication factor equal to the number of racks

bhalevy replied:
done in v2 uisng ordinal rack order.
nodes within a rack are consumed so to fill least loaded nodes first, yet nodes with equal load are shuffled.
added patch to use a tablet_id % number of racks in dc for selecting the primary replica for repair (as the first replica in the rack - hence the shuffling)

avikivity commented on test/boost/network_topology_strategy_test.cc:
```diff
@@ -468,6 +468,36 @@ SEASTAR_THREAD_TEST_CASE(NetworkTopologyStrategy_tablets_test) {
+
+    cur_tmap = std::move(tmap);
+    tmap = tab_awr_ptr->reallocate_tablets(s, stm.get(), 0, &cur_tmap).get();
+    full_ring_check(tmap, ars_ptr, stm.get());
```
It's unlikely that such a small test covers all of the complexities introduced. Check invariants and sanity in the result.

bhalevy replied:
I changed the test to randomize the cluster, number of shards, and replication options.
It covers much more now (and the current implementation in master hangs with it)

xemul commented on test/boost/network_topology_strategy_test.cc:
```diff
@@ -468,6 +468,36 @@ SEASTAR_THREAD_TEST_CASE(NetworkTopologyStrategy_tablets_test) {
+
+    cur_tmap = std::move(tmap);
+    tmap = tab_awr_ptr->reallocate_tablets(s, stm.get(), 0, &cur_tmap).get();
+    full_ring_check(tmap, ars_ptr, stm.get());
```
This amount of new tests is insufficient to cover all the corner-cases introduced by the reallocate_tablets(). It's worth thinking about either exhausting test, or randomized one with some easily-pre-evaluated invariant checking

bhalevy replied:
randomized test in v2

avikivity commented on test/boost/network_topology_strategy_test.cc:
```diff
@@ -498,6 +498,30 @@ SEASTAR_THREAD_TEST_CASE(NetworkTopologyStrategy_tablets_test) {
+
+    cur_tmap = std::move(tmap);
+    tmap = tab_awr_ptr->reallocate_tablets(s, stm.get(), 0, &cur_tmap).get();
+    full_ring_check(tmap, ars_ptr, stm.get());
```
Please check all invariants.

bhalevy replied:
I added checks about node uniqueness (also in the reallocate_tablets function so they are always checked)
what other invariants do you have in mind?

xemul commented on locator/network_topology_strategy.cc:
```diff
@@ -336,39 +335,128 @@ future<tablet_map> network_topology_strategy::allocate_tablets_for_new_table(sch
+        };
+        struct dc_rack_load {
+            size_t node_count = 0;
+            std::vector<rack_load> racks;
```
Ugh, it takes a while to understand that node_count is _not_ the amount of nodes in the racks vector :( It's worth commenting on what this fields are, or maybe splitting the struct

bhalevy replied:
yes, it's the number of nodes used in the dc.
I'll add a comment documenting that.

bhalevy replied:
removed in v2

xemul commented on locator/network_topology_strategy.cc:
```diff
@@ -336,39 +335,128 @@ future<tablet_map> network_topology_strategy::allocate_tablets_for_new_table(sch
+        // Current load per dc/rack
+        std::unordered_map<sstring, dc_rack_load> dc_rack_load_map;
+        // Current replicas per dc/rack
+        std::unordered_map<sstring, std::unordered_map<sstring, std::unordered_set<locator::host_id>>> replicas_per_dc;
```
This seems like a good candidate to fix [this](https://github.com/scylladb/scylladb/pull/17846/commits/4eccd27cb0cef090565ecab864bea46b2109aafc#r1528361930). Like this (likely not compilable, but shows the intent)

```
std::unordered_map<sstring, struct {
  size_t node_count;
  std::unordered_map<sstring, std::unordered_set<locator::host_id>> racks;
}>
```

bhalevy replied:
I can put node_count_per_dc in a separate map, for clarity

bhalevy replied:
removed in v2

bhalevy commented:
@asias please review in particular last patch in the series related to tablet repair primary replica

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7565/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7565/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/576/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8606/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7565/)
#### Build Details:
- Duration: 2 hr 6 min
- Builder: spider5.cloudius-systems.com

bhalevy commented:
One drawback of allocating replicas in racks lexicographical order is that if RF < Number of racks the first RF racks will be overloaded and we never migrate tablets between racks. In that case, we can use the replica id to start allocating in rack number `(id % number_of_racks)`.  This will still provide consistent base/view pairing if the base and view tables are allocated with the same number of initial tablets.

@nyh if and when https://github.com/scylladb/scylladb/issues/17147 is solved, would the rack order in the tablet replicas order even matter for base/view pairing, or will it filter the replicas into dc/rack groups and match the replicas in each group (I think it can use stable_sort to achieve that achieve that efficiently)

bhalevy commented:
In v3 ([bcd26c3](https://github.com/scylladb/scylladb/commit/bcd26c3118fffb54aa193dd8206d714eab812ac7)):
- rebased
- add rotation by tablet_id so that each tablets starts allocating from `rack[id % number_of_racks]`, all using lexicographically order
  - dropped "repair_tablets: distribute the primary replica across racks" as it should be needed with this approach
- network_topology_startegy_test: add NetworkTopologyStrategy_tablet_allocation_balancing_test
  - check that tablet (re)allocation is balanced across racks, nodes, shards (with a typical symmetrical topology, i.e. all same number of nodes in each rack and same number of shards in all nodes)

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7651/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7651/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/646/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8675/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7651/)
#### Build Details:
- Duration: 2 hr 17 min
- Builder: spider2.cloudius-systems.com

bhalevy commented:
@xemul, @avikivity, @tgrabiec  please review

xemul commented on test/boost/network_topology_strategy_test.cc:
```diff
@@ -480,6 +530,148 @@ SEASTAR_THREAD_TEST_CASE(NetworkTopologyStrategy_tablets_test) {
+        .with_column("v", utf8_type)
+        .build();
+
+    for (auto test_case = 0; test_case < 3; test_case++) {
```
Yikes, the "test_case" has some meaningful difference between cases, it's not an abstract "iteration with different random number(s)", right? In that case it's better to have several `NetworkTopologyStrategy_tablet_allocation_balancing_test_<something>` top-level cases calling this function with some argument that tweaks its behavior

bhalevy replied:
case 0 is rf:=1, case 2 us rf:=n and case 1, if n is large enough draws a random rf between 2 to n-1.

bhalevy replied:
in v4 I left on one random choice of rf_per_dc, and the random test function is called 100 times in a loop

xemul commented on test/boost/network_topology_strategy_test.cc:
```diff
@@ -480,6 +530,148 @@ SEASTAR_THREAD_TEST_CASE(NetworkTopologyStrategy_tablets_test) {
+    tm_cfg.topo_cfg.this_endpoint = my_address;
+    tm_cfg.topo_cfg.local_dc_rack = { snitch.local()->get_datacenter(), snitch.local()->get_rack() };
+
+    // Generate a random cluster
```
Even in debug mode this test takes 0.5 seconds to run. You use 4 random integers here:

- num_racks -- 4 values
- nodes_per_rack -- 4 values
- shard_count -- 4 values
- rf_per_dc (below) -- 13 values

Even if performing exhaustive test of _all_ values you get 4 * 4 * 4 * 13 * 0.5sec = ~6minutes. Pretty fast :)
It's worth either turning this test into exhaustive one, or at least checking ~100 "random" iterations in a loop

bhalevy replied:
with aarch64/debug it will take a year :)
I'm not sure we need need to exhaust all cases every time the test runs, as we'll get coverage over time by running this test on many occasions.

But running multiple iterations with random choices to increase coverage isn't a bad idea.

bhalevy replied:
in next version I'll refactor the tablet balancing test and call the inner function 100 times to test 100 random cases every run. That said, it makes it harder to reproduce an issue if it is generated on the Nth iteration, based on the random seed, so I'll add a printout for each call's seed so it can be reproduced this way.

bhalevy replied:
done in v4

tgrabiec commented on locator/network_topology_strategy.cc:
```diff
@@ -336,40 +337,177 @@ future<tablet_map> network_topology_strategy::allocate_tablets_for_new_table(sch
+}
+
+future<tablet_map> network_topology_strategy::reallocate_tablets(schema_ptr s, token_metadata_ptr tm, tablet_map tablets) const {
+    static thread_local std::default_random_engine rnd_engine{std::random_device{}()};
```
Unused.

bhalevy replied:
dropped in v4

tgrabiec commented on locator/network_topology_strategy.cc:
```diff
@@ -336,40 +337,177 @@ future<tablet_map> network_topology_strategy::allocate_tablets_for_new_table(sch
+        nodes.pop_back();
+        tablet_logger.trace("Allocated tablet replica={} dc={} rack={}: nodes remaining in rack={}", replica, node.dc_rack().dc, node.dc_rack().rack, nodes.size());
+        if (nodes.empty()) {
+            candidate = candidate_racks.erase(candidate);
```
shouldn't we consider new_racks here and below? There may be new_racks left, but this may jump to a non-new rack.

Maybe new_racks should be a separate list, and first we try to allocate from it, and the from the non-new list.

bhalevy replied:
done in v4

tgrabiec commented on locator/network_topology_strategy.cc:
```diff
@@ -510,6 +509,26 @@ future<tablet_replica_set> network_topology_strategy::add_tablets_in_dc(schema_p
     co_return replicas;
 }
 
+tablet_replica_set network_topology_strategy::drop_tablets_in_dc(schema_ptr s, const locator::topology& topo,
```
"s" is unused.

bhalevy replied:
The plan was to use it to print keyspace.table.  I'll add that use.

bhalevy replied:
used in v4

bhalevy commented:
In v4 ([8a77319](https://github.com/scylladb/scylladb/commit/8a77319cb723254fa6f06856935b4c49533a3b6e)):
- network_topology_strategy: allocate_tablets_for_new_table: do not rely on token ownership
  - print ks.table and tablet id in log/error messages.
- network_topology_strategy: reallocate_tablets: support deallocation via rf change
  - update load_sketch
  - print ks.table and tablet id in log/error messages.
- network_topology_startegy_test: add NetworkTopologyStrategy_tablet_allocation_balancing_test
  - call random balancing test 100 times in a loop
- network_topology_strategy: reallocate_tablets: consider new_racks before existing racks
  - addressing https://github.com/scylladb/scylladb/pull/17846#discussion_r1540232329

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7756/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7756/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8753/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/725/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7756/)
#### Build Details:
- Duration: 2 hr 23 min
- Builder: spider6.cloudius-systems.com

bhalevy commented:
@tgrabiec, @xemul please re-review.
